### PR TITLE
WIP_Update to Windows Profiler

### PR DESCRIPTION
**Pull request by batesl87**
#9224 
I've made two changes to the Profiler on Windows.
 - Randomised the time interval between samples. The interval is multiplied by a random number between 0 and 2.
- Utilised QPC to achieve microsecond sleep intervals. The use of Sleep(0) is key as it offers the rest of the time slice allocated to the Profiler thread to other threads of the same priority. 

@timholy

### CODE REVIEW

File: src/signals-win.c [hunk 1]
```diff
@@ -326,18 +326,53 @@ volatile HANDLE hBtThread = 0;
...
     // Note: illegal to use jl_* functions from this thread
-
     TIMECAPS tc;
     if (MMSYSERR_NOERROR!=timeGetDevCaps(&tc, sizeof(tc))) {
         fputs("failed to get timer resolution",stderr);
         hBtThread = 0;
         return 0;
     }
+
+	//Precision timer
```
Comment (ararslan) [inline src/signals-win.c]: Please use multiples of 4 spaces for indentation rather than tabs for consistency with the rest of the repository.

Comment (batesl87) [inline src/signals-win.c]: Sorry about that. I've recommited.

File: src/signals-win.c [hunk 2]
```diff
@@ -326,18 +326,47 @@ volatile HANDLE hBtThread = 0;
...
+            //multiply timeout(microseconds) by random factor between 0 and 2
+            DWORD timeout = (int)((nsecprof*((double)rand() / (double)RAND_MAX*2.0)/1000));
+            //if timeout is greater than a millisecond use Sleep
+            if (timeout < 1000)
+            {
+                //calculate start time for high precision sleep
+                QueryPerformanceCounter(&StartingTime);
+                int time_elapsed = 0;
+                while (time_elapsed ==0)
+                {
```
Comment (ararslan) [inline src/signals-win.c]: There's trailing whitespace on this line, which is what's causing the check to fail

**Comment by tkelman** [comment]
If we're randomizing the profile interval, we should probably do that across all platforms. Could we split the randomization from the windows-specific timer change into separate PR's?

File: src/signals-win.c [hunk 3]
```diff
@@ -326,18 +326,47 @@ volatile HANDLE hBtThread = 0;
...
-            Sleep(timeout);
+            //Convert timeout to microseconds
+            DWORD timeout = nsecprof/1000;
+            //if timeout is greater than a millisecond use Sleep
+            if (timeout < 1000)
+            {
+                //calculate start time for high precision sleep
+                QueryPerformanceCounter(&StartingTime);
+                int time_elapsed = 0;
+                while (time_elapsed ==0)
```
Comment (tkelman) [inline src/signals-win.c]: very minor, but would look a little nicer with spaces on both sides of the `==`, not just on the left

File: src/signals-win.c [hunk 4]
```diff
@@ -326,18 +326,47 @@ volatile HANDLE hBtThread = 0;
...
+                    ElapsedMicroseconds.QuadPart = EndingTime.QuadPart - StartingTime.QuadPart;
+                    ElapsedMicroseconds.QuadPart *= 1000000;
+                    ElapsedMicroseconds.QuadPart /= Frequency.QuadPart;
+                    if (ElapsedMicroseconds.QuadPart >= timeout)
+                    {
+                        time_elapsed = 1;
+                    }
+                    else
+                    {
+                        //if time has not elasped offer up processor to another thread
```
Comment (tkelman) [inline src/signals-win.c]: elapsed

**Comment by tkelman** [comment]
for reference/reviewer selection, looks like this code was originally authored in https://github.com/JuliaLang/julia/commit/8247ac7321ea195dac7358902169234bb12fbcbc

**Comment by tkelman** [comment]
is this still WIP, or RFC?

**Comment by batesl87** [comment]
It's still WIP. Small profiler intervals can cause julia to crash (see below). I haven't had a chance to chase it down, but could this be due to the StackWalk64 not being thread-safe as mentioned in [https://github.com/JuliaLang/julia/issues/2597](url) @vtjnash

Please submit a bug report with steps to reproduce this fault, and any error messages that follow (in their entirety). Thanks.
Exception: EXCEPTION_ACCESS_VIOLATION at 0x7ff8ed36bc62 -- RtlVirtualUnwind at C:\WINDOWS\SYSTEM32\ntdll.dll (unknown line)
while loading no file, in expression starting on line 0

**Comment by vtjnash** [comment]
There's still interest in #9224, but we should do it on all platforms. We don't want to busy wait (like this PR), since that can skew the results even more significantly however.

**Pull request closed without merging**